pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -2055,28 +2085,42 @@ func (c *Cloud) GetDiskPath(volumeName KubernetesVolumeID) (string, error) {
 
 // DiskIsAttached implements Volumes.DiskIsAttached
 func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeName) (bool, error) {
-	_, instance, err := c.getFullInstance(nodeName)
-	if err != nil {
-		if err == cloudprovider.InstanceNotFound {
-			// If instance no longer exists, safe to assume volume is not attached.
-			glog.Warningf(
-				"Instance %q does not exist. DiskIsAttached will assume disk %q is not attached to it.",
-				nodeName,
-				diskName)
-			return false, nil
-		}
+	disk, err := newAWSDisk(c, diskName)
 
-		return false, err
+	if err != nil {
+		// We will assume disk is still attached just to be safe
+		return true, err
 	}
 
-	diskID, err := diskName.mapToAWSVolumeID()
+	info, err := disk.describeVolume()
+
 	if err != nil {
-		return false, fmt.Errorf("error mapping volume spec %q to aws id: %v", diskName, err)
+		glog.Errorf("Error describing volume %q: %q", disk.awsID, err)
+		// if for some reason we can not describe volume we will return error and assume disk is detached
+		// because it is probably deleted
+		return false, err
 	}
 
-	for _, blockDevice := range instance.BlockDeviceMappings {
-		id := awsVolumeID(aws.StringValue(blockDevice.Ebs.VolumeId))
-		if id == diskID {
+	volumeState := aws.StringValue(info.State)
+
+	// If volume is available it means it is detached
+	if volumeState == "available" {
+		return false, nil
+	}
+
+	if len(info.Attachments) > 0 {
+		attachment := info.Attachments[0]
+		instanceID := aws.StringValue(attachment.InstanceId)
+		attachedInstance, ierr := c.getInstanceByID(instanceID)
+
+		// This should never happen but if it does it could mean there was a race and instance
+		// has been deleted, in which case volume is detached anyways.
+		if ierr != nil {
+			glog.Errorf("Error fetching instance %s for volume %s", instanceID, diskName)
+			return false, nil
+		}
+		attachedNodeName := mapInstanceToNodeName(attachedInstance)
+		if attachedNodeName == nodeName {
```

Comment:
a lot of this code is duplicated above in `DetachDisk`. Would it make sense to put this common code in a helper func?

Reply:
removed duplication